lock-api/src/main/java/com/palantir/lock/watch/LockWatchEvent.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+public interface LockWatchEvent {
```

Comment:
it would look like this:

```java
@Data
abstract class LockWatchEvent {
  interface EventTypes<R> {
    R lockWatchCreatedEvent(long sequence, LockWatchRequest request);
    R lockEvent(long sequence, LockToken lockToken, Set<LockDescriptor> lockDescriptors);
    R unlockEvent(long sequence, LockToken lockToken, Set<LockDescriptor> lockDescriptors);
  }

  public abstract <R> R match(EventTypes<R> eventTypes);

  @Override
  public abstract int hashCode();

  @Override
  public abstract boolean equals(Object obj);

  @Override
  public abstract String toString();
}
```

It generates you all the different classes, and gives you convenient accessors, although you can just define methods here if you need as well.

Reply:
From `LeaderPingResult`, here's an example usage:

```java
    public boolean isSuccessful() {
        return LeaderPingResults.caseOf(this)
                .pingReturnedTrue_(true)
                .otherwise_(false);
    }
```

It has useful lens accessors e.g. for `LeaderPingResult` 
```java
static Optional<Throwable> getException(LeaderPingResult leaderPingResult);
static Function<LeaderPingResult, LeaderPingResult> setException(Throwable newException);
static Function<LeaderPingResult, LeaderPingResult> modException(
      Function<Throwable, Throwable> exceptionMod);
```

`getException` is optional because it's not on each case, but if an attribute *is* on each case, then you get the unwrapped value, lens ftw.

Reply:
reference material: https://github.com/derive4j/derive4j

Reply:
don't think there's much value from this